apps/aecore/src/aec_chain_state.erl

```diff
@@ -247,70 +257,94 @@ internal_insert(Node, Block) ->
-assert_previous_height(#node{type = NodeType} = Node) ->
-    case {NodeType, db_find_node(prev_hash(Node))} of
-        {key, {ok, PrevNode}} ->
+assert_previous_height(Node) ->
+    case is_key_block(Node) of
+        true ->
+            {ok, PrevNode} = db_find_node(prev_hash(Node)),
             case node_height(PrevNode) =:= (node_height(Node) - 1) of
                 true -> ok;
                 false -> internal_error(height_inconsistent_for_keyblock_with_previous_hash)
             end;
-        {micro, {ok, PrevNode}} ->
+        false ->
+            {ok, PrevNode} = db_find_node(prev_hash(Node)),
             case node_height(PrevNode) =:= node_height(Node) of
                 true -> ok;
                 false -> internal_error(height_inconsistent_for_microblock_with_previous_hash)
-            end;
-        {_, error} -> ok
+            end
     end.
 
-%% To assert the target calculation we need DeltaHeight headers counted
-%% backwards from the node we want to assert. If Height <= DeltaHeight
-%% we will need all headers back to genesis.
-%% NG-INFO: micro blocks are signed, so we don't validate the target
-assert_calculated_target(#node{type = micro}) -> ok;
+%% To assert key block target calculation we need DeltaHeight headers counted
+%% backwards from the node we want to assert.
 assert_calculated_target(Node) ->
+    case is_key_block(Node) of
+        true  -> assert_key_block_target(Node);
+        false -> ok
+    end.
+
+assert_key_block_target(Node) ->
     case db_find_node(prev_hash(Node)) of
         error -> ok;
         {ok, PrevNode} ->
-            case node_height(Node) of
-                0  -> ok;
-                Height ->
-                    Delta = aec_governance:blocks_to_check_difficulty_count(),
-                    assert_calculated_target(Node, PrevNode, Delta, Height)
+            Delta         = aec_governance:key_blocks_to_check_difficulty_count(),
+            Height        = node_height(Node),
+            GenesisHeight = aec_block_genesis:genesis_header(),
+            case Delta >= Height - GenesisHeight of
+                true ->
+                    %% We only need to verify that the target is equal to its predecessor.
+                    assert_difficulty_equal_to_prev(Node, PrevNode);
+                false ->
+                    assert_calculated_target(Node, PrevNode, Delta)
             end
     end.
 
-assert_calculated_target(Node, PrevNode, Delta, Height) when Delta >= Height ->
-    %% We only need to verify that the target is equal to its predecessor.
-    case {node_difficulty(Node), node_difficulty(PrevNode)} of
+assert_difficulty_equal_to_prev(Node, PrevNode) ->
+    PrevKeyNode = case is_key_block(PrevNode) of
+                      true  -> PrevNode;
+                      false -> db_find_node(node_key_hash(PrevNode))
+                  end,
+    case {node_difficulty(Node), node_difficulty(PrevKeyNode)} of
```

Comment:
We should probably compare targets instead of difficulties to avoid comparing floats for equality.

Reply:
Good point.